### PR TITLE
FIX LOG IN LOOP

### PR DESCRIPTION
**Pull request by dinusharath**
FIX LOG IN LOOP
stored reseller token in databse and retrieved as discussed.use db instead of session to get token 

https://openprovider.atlassian.net/browse/INTGRTNS-175
https://openprovider.atlassian.net/browse/INTGRTNS-176

### CODE REVIEW

File: modules/registrars/openprovider/Controllers/System/ConfigController.php [hunk 1]
```diff
@@ -179,7 +182,15 @@ protected function checkCredentials($configarray, $params)
             Configuration::get('api_url') :
             Configuration::get('api_url_cte');
 
-        if ($this->session->has('AUTH_TOKEN')) {
+ $token_result = [];
```
Comment (dmirou) [inline modules/registrars/openprovider/Controllers/System/ConfigController.php:185]: Please reformat the code

File: modules/registrars/openprovider/Controllers/System/ConfigController.php [hunk 2]
```diff
@@ -179,7 +182,15 @@ protected function checkCredentials($configarray, $params)
...
+
+        if (Capsule::schema()->hasTable('reseller_tokens')) {
+            $token_result = Capsule::table('reseller_tokens')->where('username', $params['Username'])->orderBy('created_at', 'desc')->get();
```
Comment (dmirou) [inline modules/registrars/openprovider/Controllers/System/ConfigController.php:188]: Maybe it's better to add limit 1?

File: modules/registrars/openprovider/Controllers/System/ConfigController.php [hunk 3]
```diff
@@ -179,7 +182,15 @@ protected function checkCredentials($configarray, $params)
...
+        } 
+
+        $expireTime = count($token_result) > 0 ? new Carbon($token_result[0]->expire_at) : null;
+
+        if (count($token_result) > 0 && Carbon::now()->diffInSeconds($expireTime, false) > 0) {
```
Comment (dmirou) [inline modules/registrars/openprovider/Controllers/System/ConfigController.php:193]: if ($expireTime && ...
expirteTime will be null if count($token_result) <= 0

File: modules/registrars/openprovider/init.php [hunk 4]
```diff
@@ -77,15 +80,53 @@ function openprovider_bind_required_classes($launcher)
         $client = new ApiV1($logger, $camelCaseToSnakeCaseNameConverter, $idn);
         $client->getConfiguration()->setHost($host);
 
-        if (!$session->has('AUTH_TOKEN')) {
+        $token_result = [];
+
+        if (Capsule::schema()->hasTable('reseller_tokens')) {
+            $token_result = Capsule::table('reseller_tokens')->where('username', $params['Username'])->orderBy('created_at', 'desc')->get();
```
Comment (dmirou) [inline modules/registrars/openprovider/init.php:86]: the same as above, var name, limit

File: modules/registrars/openprovider/init.php [hunk 5]
```diff
@@ -77,15 +80,53 @@ function openprovider_bind_required_classes($launcher)
...
+                    $table->string('expire_at');
+                    $table->string('created_at');
+                }
+            );
+        }
+
+        $token = "";
+        $expireTime = count($token_result) > 0 ? new Carbon($token_result[0]->expire_at) : null;
+
+        if (count($token_result) > 0 && Carbon::now()->diffInSeconds($expireTime, false) > 0) {
```
Comment (dmirou) [inline modules/registrars/openprovider/init.php:104]: the same as above

File: modules/registrars/openprovider/init.php [hunk 6]
```diff
@@ -77,15 +80,53 @@ function openprovider_bind_required_classes($launcher)
...
+
+            Capsule::table('reseller_tokens')->where('username', $params['Username'])->delete();
+
+            Capsule::connection()->transaction(
+                function ($connectionManager) use ($token, $params) {
+                    $connectionManager->table('reseller_tokens')->insert(
+                        [
+                            'username' => $params['Username'],
+                            'token' => $token,
+                            'expire_at' => Carbon::now()->addDays(2)->toDateTimeString(),
```
Comment (dmirou) [inline modules/registrars/openprovider/init.php:120]: Move 2 to class const

File: modules/registrars/openprovider/init.php [hunk 7]
```diff
@@ -77,15 +80,53 @@ function openprovider_bind_required_classes($launcher)
...
+        } else {
             $token = $client->call('generateAuthTokenRequest', [
                 'username' => $params['Username'],
                 'password' => $params['Password']
             ])->getData()['token'];
-            $session->set('AUTH_TOKEN', $token);
+
+            Capsule::table('reseller_tokens')->where('username', $params['Username'])->delete();
+
+            Capsule::connection()->transaction(
```
Comment (dmirou) [inline modules/registrars/openprovider/init.php:114]: why do you need transaction here?
Is there a few operations inside insert?

**Comment by dinusharath** [comment]
@dmirou  i have fixed these. thanks a lot for the feedbacks.

File: modules/registrars/openprovider/constants/Constants.php [hunk 8]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+namespace OpenProvider\WhmcsRegistrar\Constants;
+
+use Carbon\Carbon;
+
+class Constants
+{
+    public static function getAuthTokenExpirationTimeFromNow()
```
Comment (dmirou) [inline modules/registrars/openprovider/constants/Constants.php:9]: No, I mean const like 
`const SESSION_EXPIRATION_LIFE_TIME = 300;`
In openprovider/init.php

File: modules/registrars/openprovider/Controllers/System/ConfigController.php [hunk 9]
```diff
@@ -179,7 +181,17 @@ protected function checkCredentials($configarray, $params)
...
-        if ($this->session->has('AUTH_TOKEN')) {
+
+        $tokenResult = null;
+
+        if (Capsule::schema()->hasTable('reseller_tokens')) {
+            $tokenResult = Capsule::table('reseller_tokens')->where('username', $params['Username'])->orderBy('created_at', 'desc')->first();
+        }
+
+        $expireTime = is_null($tokenResult) ? false : new Carbon($tokenResult->expire_at);
+        $isExpired = $expireTime ? Carbon::now()->diffInSeconds($expireTime, false) < 0 : true;
```
Comment (dmirou) [inline modules/registrars/openprovider/Controllers/System/ConfigController.php:192]: I have an idea how to do it better:
$isAlive = $expireTime && Carbon::now()->diffInSeconds($expireTime, false) < 0;
if ($tokenResult && $isAlive) {
...

File: modules/registrars/openprovider/init.php [hunk 10]
```diff
@@ -77,15 +81,48 @@ function openprovider_bind_required_classes($launcher)
...
+                    $table->string('token');
+                    $table->string('expire_at');
+                    $table->string('created_at');
+                }
+            );
+        }
+
+        $token = "";
+        $expireTime = is_null($tokenResult) ? false : new Carbon($tokenResult->expire_at);
+        $isExpired = $expireTime ? Carbon::now()->diffInSeconds($expireTime, false) < 0 : true;
```
Comment (dmirou) [inline modules/registrars/openprovider/init.php:104]: The same here:
$isAlive = $expireTime && Carbon::now()->diffInSeconds($expireTime, false) < 0;
if ($tokenResult && $isAlive) {
...

File: modules/registrars/openprovider/init.php [hunk 11]
```diff
@@ -77,15 +81,48 @@ function openprovider_bind_required_classes($launcher)
...
+
+        if (!is_null($tokenResult) && $expireTime && !$isExpired) {
```
Comment (dmirou) [inline modules/registrars/openprovider/init.php:106]: move $token = ""; right above if

**Review by dmirou — APPROVED** [review]
well done :+1:

**Pull request merged**